src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs

```diff
@@ -640,7 +640,16 @@ public CompilationInfo(Compilation compilation, bool hasSuccessfullyLoaded)
                                 if (metadataReference != null)
                                 {
                                     newReferences.Add(metadataReference);
-                                    if (!metadataReferenceToProjectId.ContainsKey(metadataReference))
+                                    if (metadataReferenceToProjectId.TryGetValue(metadataReference, out var alreadyReferencedProjectId))
```

Comment:
Having just written that, I found some code (Debug.Assert in Microsoft.CodeAnalysis.ProjectState:500 ) is trying to prevent duplicates slipping through. Should we instead just make error message there helpful?

Reply:
Ah dammit, sorry forgot to communicate this back to you as well. So I dug a bit further and chatted with the team here, and we decided that yes, we should just fix the API to properly block the duplicates. That's an old bug we've tracked for awhile (#12101) and so yes, we'll want to just fix that. I reassigned that bug back to me but haven't done anything yet so feel free to give a shot at it if you'd like. In the meantime, our conclusion was the primary way people were hitting this was using tools that use MSBuildWorkspace, so I directly wrote a fix to prevent the duplicate entries in the first place: #33262.  Obviously that fix would need to go in first before a fix for #12101 because that'd just be making the crash even worse.

Our concern was whether or not fixing the block becomes a back compat issue: code out there is getting an exception for something that previously "worked". Since you observed that it'll still crash here we're not too worried that it's a problem.

Reply:
So thanks for this PR since it helped us all confirm we understood the issue, but at this point we won't take a change *here*. If you want to tackle #12101, go ahead!

Reply:
Yep that's the right thing to do I think. Cheers @jasonmalinowski